crates/subspace-fraud-proof/src/verifier_api.rs

```diff
@@ -0,0 +1,176 @@
+//! This module derives an trait [`VerifierApi`] from the runtime api `ReceiptsApi`
+//! as well as the implementation to provide convenient interfaces used in the fraud
+//! proof verification.
+
+use codec::{Decode, Encode};
+use sc_client_api::HeaderBackend;
+use sp_api::ProvideRuntimeApi;
+use sp_core::H256;
+use sp_domains::fraud_proof::{ExecutionPhase, InvalidStateTransitionProof, VerificationError};
+use sp_domains::DomainId;
+use sp_receipts::ReceiptsApi;
+use sp_runtime::traits::{Block as BlockT, NumberFor};
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+/// This trait abstracts convenient APIs for the fraud proof verifier.
+pub trait VerifierApi {
+    /// Verifies whether `pre_state_root` declared in the proof is same as the one recorded on chain.
+    fn verify_pre_state_root(
+        &self,
+        invalid_state_transition_proof: &InvalidStateTransitionProof,
+    ) -> Result<(), VerificationError>;
+
+    /// Verifies whether `post_state_root` declared in the proof is different from the one recorded on chain.
+    fn verify_post_state_root(
+        &self,
+        invalid_state_transition_proof: &InvalidStateTransitionProof,
+    ) -> Result<(), VerificationError>;
+
+    /// Returns the hash of primary block at height `domain_block_number`.
+    fn primary_hash(
+        &self,
+        domain_id: DomainId,
+        domain_block_number: u32,
+    ) -> Result<H256, VerificationError>;
```

Comment:
Perhaps moving `primary_hash` and the incoming `state_root` to a dedicated trait will be nicer, I remember @vedhavyas had a similar suggestion when introducing `primary_hash` to `VerifyPrePostStateRoot`.